core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -598,6 +603,86 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
     }, "Broker fail to initialize after restart")
   }
 
+  @Test
+  def testPreemptionOnControllerShutdown(): Unit = {
+    servers = makeServers(1, enableControlledShutdown = false)
+    val controller = getController().kafkaController
+    val count = new AtomicInteger(2)
+    val latch = new CountDownLatch(1)
+    val spyThread = spy(controller.eventManager.thread)
+    controller.eventManager.setControllerEventThread(spyThread)
+    val processedEvent = new MockEvent(ControllerState.TopicChange) {
+      override def process(): Unit = latch.await()
+      override def preempt(): Unit = {}
+    }
+    val preemptedEvent = new MockEvent(ControllerState.TopicChange) {
+      override def process(): Unit = {}
+      override def preempt(): Unit = count.decrementAndGet()
```

Comment:
This method should be executed by the thread that is running this test. If you agree, no need to use an `AtomicInteger`.